docs/how-to/hip_runtime_api/multi_device.rst

```diff
@@ -0,0 +1,423 @@
+
+In a multi-device system, streams and events are essential for efficient
+parallel computation and synchronization. Streams enable asynchronous task
+execution, allowing multiple devices to process data concurrently without
+blocking one another. Events provide a mechanism for synchronizing operations
+across devices, ensuring that tasks on one device are completed before
+dependent tasks on another device begin. This coordination prevents race
+conditions and optimizes data flow in multi-GPU systems. Together, streams and
+events maximize performance by enabling parallel execution, load balancing, and
+effective resource utilization across heterogeneous hardware.
+
+.. code-block:: cpp
+
+    #include <hip/hip_runtime.h>
+    #include <iostream>
+
+    __global__ void simpleKernel(double *data)
+    {
+        int idx = blockIdx.x * blockDim.x + threadIdx.x;
+        data[idx] = idx * 2.0;
+    }
+
+    int main()
+    {
+        int numDevices;
+        hipGetDeviceCount(&numDevices);
+
+        if (numDevices < 2) {
+            std::cerr << "This example requires at least two GPUs." << std::endl;
+            return -1;
+        }
+
+        double *deviceData0, *deviceData1;
+        size_t size = 1024 * sizeof(*deviceData0);
+
+        // Create streams and events for each device
+        hipStream_t stream0, stream1;
+        hipEvent_t startEvent0, stopEvent0, startEvent1, stopEvent1;
+
+        // Initialize device 0
+        hipSetDevice(0);
+        hipStreamCreate(&stream0);
+        hipEventCreate(&startEvent0);
+        hipEventCreate(&stopEvent0);
+        hipMalloc(&deviceData0, size);
+
+        // Initialize device 1
+        hipSetDevice(1);
+        hipStreamCreate(&stream1);
+        hipEventCreate(&startEvent1);
+        hipEventCreate(&stopEvent1);
+        hipMalloc(&deviceData1, size);
+
+        // Record the start event on device 0
+        hipSetDevice(0);
+        hipEventRecord(startEvent0, stream0);
+
+        // Launch the kernel asynchronously on device 0
+        simpleKernel<<<1000, 128, 0, stream0>>>(deviceData0);
+
+        // Record the stop event on device 0
+        hipEventRecord(stopEvent0, stream0);
+
+        // Wait for the stop event on device 0 to complete
+        hipEventSynchronize(stopEvent0);
+
+        // Record the start event on device 1
+        hipSetDevice(1);
+        hipEventRecord(startEvent1, stream1);
+
+        // Launch the kernel asynchronously on device 1
+        simpleKernel<<<1000, 128, 0, stream1>>>(deviceData1);
+
+        // Record the stop event on device 1
+        hipEventRecord(stopEvent1, stream1);
+
+        // Wait for the stop event on device 1 to complete
+        hipEventSynchronize(stopEvent1);
+
+        // Calculate elapsed time between the events for both devices
+        float milliseconds0 = 0, milliseconds1 = 0;
+        hipEventElapsedTime(&milliseconds0, startEvent0, stopEvent0);
```

Comment:
Presumably this is where the host code waits?

Reply:
hipEventElapsedTime is a non-blocking operation, so we need synchronization beforehand to ensure accurate timing measurements.